common/membership/hashring.go

```diff
@@ -134,40 +147,41 @@ func (r *ringpopHashring) Stop() {
 }
 
 // Lookup finds the host in the ring responsible for serving the given key
-func (r *ringpopHashring) Lookup(
+func (r *ring) Lookup(
 	key string,
 ) (*HostInfo, error) {
-
 	addr, found := r.ring().Lookup(key)
 	if !found {
 		select {
-		case r.refreshChan <- struct{}{}:
+		case r.refreshChan <- &ChangedEvent{}:
 		default:
 		}
 		return nil, ErrInsufficientHosts
 	}
-	return NewHostInfo(addr, r.getLabelsMap()), nil
+	return NewHostInfo(addr), nil
 }
 
-func (r *ringpopHashring) AddListener(
-	name string,
+// Subscribe registers callback watcher.
+// All subscribers are notified about ring change.
+func (r *ring) Subscribe(
+	service string,
 	notifyChannel chan<- *ChangedEvent,
 ) error {
-
 	r.smu.Lock()
 	defer r.smu.Unlock()
-	_, ok := r.subscribers[name]
+	_, ok := r.subscribers[service]
 	if ok {
-		return fmt.Errorf("listener already exist for service %q", name)
+		return fmt.Errorf("service %q already subscribed", service)
 	}
-	r.subscribers[name] = notifyChannel
+
+	r.subscribers[service] = notifyChannel
 	return nil
 }
 
-func (r *ringpopHashring) RemoveListener(
+// Unsubscribe removes subscriber
+func (r *ring) Unsubscribe(
 	name string,
 ) error {
-
 	r.smu.Lock()
 	defer r.smu.Unlock()
 	_, ok := r.subscribers[name]
```

Comment:
it's just nearby code, but: delete's safe to do on non-existent keys, this check could be skipped.

Reply:
done